### PR TITLE
'basestring' fix applied

### PR DESCRIPTION
**Pull request by vilnius-dev**
basestring fix applied to WMCore/src/python folder, converts **basestring** to **str**

Example:

python2
-        if isinstance(status, basestring):
python3
+        if isinstance(status, str):

### CODE REVIEW

**Comment by ericvaandering** [comment]
I think this is going to blow up, unfortunately. In python2 basestr also includes unicode while str does not.

**Comment by vkuznet** [comment]
If you'll add my name explicitly I'll comment earlier. Eric is right, this PR will break the code since basestring covers both str and unicode. Therefore you (Roman) need to adjust code accordingly and change every statement to cover both use cases, e.g.
```
if (isinstance(status, str) or isinstance(status, unicode))
```
Please be aware that both conditions should be applied through OR operator and every if or if not should cover them accordingly.

**Comment by ericvaandering** [comment]
Can be a bit simpler. Is instance takes a tuple as the 2nd parameter

Sent from a mobile device. Please excuse my brevity or transcription errors.

On Aug 7, 2017, at 10:53, Valentin Kuznetsov <notifications@github.com<mailto:notifications@github.com>> wrote:


If you'll add my name explicitly I'll comment earlier. Eric is right, this PR will break the code since basestring covers both str and unicode. Therefore you (Roman) need to adjust code accordingly and change every statement to cover both use cases, e.g.

if (isinstance(status, str) or isinstance(status, unicode))


Please be aware that both conditions should be applied through OR operator and every if or if not should cover them accordingly.

**Comment by amaltaro** [comment]
Sorry for creating confusion, but I though all strings were unicode in python3 (all under the str class/data type), thus there is no unicode type. Am I wrong?

File: src/python/WMCore/Wrappers/JsonWrapper/JSONThunker.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 
     # Make types compatible, can also be removed once transition is over
     long = int
-    basestring = str
+    str = str
```
Comment (vkuznet) [inline src/python/WMCore/Wrappers/JsonWrapper/JSONThunker.py]: this is ambiguous and should not be defined

File: src/python/WMCore/Wrappers/JsonWrapper/JSONThunker.py [hunk 2]
```diff
@@ -46,7 +46,7 @@ def __init__(self):
                                  str,
                                  bytes,
                                  unicode,
```
Comment (vkuznet) [inline src/python/WMCore/Wrappers/JsonWrapper/JSONThunker.py]: unicode is not defined in python3 therefore should be removed manually. At the same time bytes is not strings anymore. Therefore I would review this class manually.

**Review by vkuznet — COMMENTED** [review]
My fault, in python3 unicode and str are the same, so we need to test another way around, i.e.
if isinstance(s, unicode) should be converted to if isinstance(s, str). I also added couple of comments to code which should be revisited manually. I would suggest to write simple stand-alone code and test its logic in python2 and python3.

**Comment by ericvaandering** [comment]
That's why I said initially that this fix only makes sense also when doing another one, which I think is "builtins". That adds a "str" class to python2 that is supposed to work like str in python3.  That mostly works, I think, but causes some problems

But look, we have to temper our expectations. We are not going to have code that really works on python2 and python3, I think. But what we can work towards is code that mostly does that and when we can really test things on python3, we can do an automated conversion of the last little bit and maintain both code bases that way. At least that is my thought. I don't expect we need to maintain both code bases for long when we finally decide to go for it.

**Comment by vkuznet** [comment]
Eric,
we can do slightly differently. Replace
```if instance(s, basestring)```
into
```
from six import string_types
if instance(s, string_types)
```
which defines string_types differently in py2/py3.
In py2 string_types is a ```(<type 'basestring'>,)``` while in py3 it is ```(<class 'str'>,)```.
The problem is that in py2 we still need to cover unicode.

**Comment by ericvaandering** [comment]
No. No six. I'm open to version defined if statements if that is the last option.

Sent from a mobile device. Please excuse my brevity or transcription errors.

On Aug 8, 2017, at 04:02, Valentin Kuznetsov <notifications@github.com<mailto:notifications@github.com>> wrote:

Eric,
we can do slightly differently. Replace
```if instance(s, basestring)```
into
```
from six import string_types
if instance(s, string_types)
```
which defines string_types differently in py2/py3.
In py2 string_types is a ```(<type 'basestring'>,)``` while in py3 it is ```(<class 'str'>,)```.
The problem is that in py2 we still need to cover unicode.

**Comment by ericvaandering** [comment]
But moreover, I'm going to stick to my original recommendation. The str/basestring/unicode mess is probably the hardest thing to do during this transition, so I think it's best to forget about or postpone this until the rest is in good shape. 

I sent @amaltaro and @vkuznet some ideas about the lowest hanging fruit. I'd suggest starting there (or maybe @vkuznet has other ideas). But I don't think this is the place to start with our code.

**Comment by amaltaro** [comment]
I couldn't agree more. Honestly, if the py2 to py3 migration is straight forward (as it would be for this feature) AND the futurize procedure is not (or depends on third -party libraries), I'm all in favor of keep a bookkeeping of those changes (maybe a GH issue flagged as python3, with the exact change and/or command needed for the migration) and forgot it for now until we get to the point where we're about to migrate to python3.

Otherwise we may increase the dependencies, code performance, etc for a couple of years "just" for the sake of having a hybrid code that runs on py2 and py3. Just my 2 cents :)

**Comment by mapellidario** [comment]
> The str/basestring/unicode mess is probably the hardest thing to do during this transition, so I think it's best to forget about or postpone this until the rest is in good shape.

I could not agree more! However, we decided not to split the work between easy (first) and hard (later), but with the gradual migration we deal with all changes (easy and hard) together only on a portion of dmwm/WMCore at a time.
In particular. dealing with `basestring` is part of our current efforts with string in py2 and py3 and it is taken care of. 

I would suggest to close this PR without merging

**Comment by amaltaro** [comment]
Agreed! Thanks Roman and Dario for your effort! Closing it

**Pull request closed without merging**